File: src/com/timepath/plaf/x/filechooser/SwingFileChooser.java
package com.timepath.plaf.x.filechooser;

import com.timepath.plaf.OS;
import java.io.File;
import java.util.logging.Logger;
import javax.swing.JFileChooser;
import javax.swing.filechooser.FileFilter;

/**
 *
 * @author timepath
 */
public class SwingFileChooser extends BaseFileChooser {

    @Override
    public File[] choose() {
        if(OS.isLinux()) {
//            UIManager.put("FileChooserUI", "eu.kostia.gtkjfilechooser.ui.GtkFileChooserUI");
        }
        JFileChooser fd = new JFileChooser(directory);
        fd.setAcceptAllFileFilterUsed(true);
        for(final ExtensionFilter ef : filters) {
            FileFilter ff = new FileFilter() {
                public boolean accept(File file) {
                    for(String e : ef.getExtensions()) {
                        if(file.getName().matches(".+" + e)) {
                            return true;
                        }
                    }
                    return file.isDirectory();
                }

                public String getDescription() {
                    StringBuilder filter = new StringBuilder();
                    filter.append(" (*").append(ef.getExtensions().get(0));
                    for(String e : ef.getExtensions().subList(1, ef.getExtensions().size())) {
                        filter.append(", *").append(e);
                    }
                    filter.append(")");
                    return ef.getDescription() + filter.toString();
                }
            };
            fd.addChoosableFileFilter(ff);
            if(fd.isAcceptAllFileFilterUsed()) {
                fd.setAcceptAllFileFilterUsed(false);
                fd.setFileFilter(ff);
            }
        }
        fd.setDialogTitle(dialogTitle);
        fd.setDialogType(this.isSaveDialog() ? JFileChooser.SAVE_DIALOG : JFileChooser.OPEN_DIALOG);
        if(directory != null) {
            if(file != null) {
                fd.setSelectedFile(new File(directory, file));
            } else {
                fd.setSelectedFile(directory);
            }
        }
        fd.setFileSelectionMode(this.isDirectoryMode() ? JFileChooser.DIRECTORIES_ONLY : JFileChooser.FILES_AND_DIRECTORIES);
        if(!this.isSaveDialog()) {
            fd.setMultiSelectionEnabled(multiSelectionEnabled);
        }
        File[] selection = null;
        if(fd.showDialog(parent, approveButtonText) == JFileChooser.APPROVE_OPTION) {
            if(this.multiSelectionEnabled) {
                selection = fd.getSelectedFiles();
            } else {
                selection = new File[]{fd.getSelectedFile()};
            }
        }
        return selection;
    }

    private static final Logger LOG = Logger.getLogger(SwingFileChooser.class.getName());

}

File: src/main/kotlin/com/timepath/plaf/x/filechooser/SwingFileChooser.kt
package com.timepath.plaf.x.filechooser

import com.timepath.plaf.OS
import java.io.File
import java.util.logging.Logger
import javax.swing.JFileChooser
import javax.swing.filechooser.FileFilter

/**
 * @author TimePath
 */
public class SwingFileChooser : BaseFileChooser() {

    override fun choose(): Array<File>? {
        if (OS.isLinux()) {
            //            UIManager.put("FileChooserUI", "eu.kostia.gtkjfilechooser.ui.GtkFileChooserUI");
        }
        val fd = JFileChooser()
        fd.setAcceptAllFileFilterUsed(true)
        for (ef in filters) {
            val fileFilter = object : FileFilter() {
                override fun accept(f: File): Boolean {
                    for (e in ef.getExtensions()) {
                        if (f.getName().matches(".+" + e)) {
                            return true
                        }
                    }
                    return f.isDirectory()
                }

                override fun getDescription(): String {
                    val filter = StringBuilder()
                    filter.append(" (*").append(ef.getExtensions().get(0))
                    for (e in ef.getExtensions().subList(1, ef.getExtensions().size())) {
                        filter.append(", *").append(e)
                    }
                    filter.append(')')
                    return ef.description + filter
                }
            }
            fd.addChoosableFileFilter(fileFilter)
            if (fd.isAcceptAllFileFilterUsed()) {
                fd.setAcceptAllFileFilterUsed(false)
                fd.setFileFilter(fileFilter)
            }
        }
        fd.setDialogTitle(dialogTitle)
        fd.setDialogType(if (isSaveDialog()) JFileChooser.SAVE_DIALOG else JFileChooser.OPEN_DIALOG)
        if (directory != null) {
            val fileName = fileName
            if (fileName != null) {
                fd.setSelectedFile(File(directory, fileName))
            } else {
                fd.setCurrentDirectory(directory)
            }
        }
        fd.setFileSelectionMode(if (isDirectoryMode()) JFileChooser.DIRECTORIES_ONLY else JFileChooser.FILES_AND_DIRECTORIES)
        if (!isSaveDialog()) {
            fd.setMultiSelectionEnabled(multiSelectionEnabled)
        }
        var selection: Array<File>? = null
        if (fd.showDialog(parent, approveButtonText) == JFileChooser.APPROVE_OPTION) {
            selection = if (multiSelectionEnabled) fd.getSelectedFiles() else array(fd.getSelectedFile())
        }
        return selection
    }

    companion object {

        private val LOG = Logger.getLogger(javaClass<SwingFileChooser>().getName())
    }
}
